File: LeetCode/215.Kth_Largest_Element_in_an_Array/215.Kth_Largest_Element_in_an_Array.py
#!/usr/bin/env python3


# heap
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        heap = []
        for num in nums:
            heapq.heappush(heap, num)
            if len(heap) > k:
                heapq.heappop(heap)
        return heapq.heappop(heap)


# divide and conquer
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:

        # divide and conquer
        def quick_select(l, r, i):
            if l == r:
                return nums[l]
            pivot_index = random.randint(l, r)
            nums[pivot_index], nums[r] = nums[r], nums[pivot_index]
            pivot = nums[r]

            # check num is smaller than pivot, then switch to left side
            j = l
            for p in range(l, r):
                if nums[p] < pivot:
                    nums[p], nums[j] = nums[j], nums[p]
                    j += 1
            # put pivot val in right place jth index
            nums[r], nums[j] = nums[j], nums[r]
            if j == i:
                return nums[j]
            elif j < i:
                return quick_select(j + 1, r, i)
            else:
                return quick_select(l, j - 1, i)

        n = len(nums)
        return quick_select(0, n - 1, n - k)

File: LeetCode/215.Kth_Largest_Element_in_an_Array/215.Kth_Largest_Element_in_an_Array.swift
// heap
class Solution {
    func findKthLargest(_ nums: [Int], _ k: Int) -> Int {
        var heap = Heap<Int>()

        for num in nums {
            heap.insert(num)
            if heap.count > k {
                heap.popMin()
            }
        }

        return heap.popMin()!
    }
}

// quick select
class Solution {
    func findKthLargest(_ nums: [Int], _ k: Int) -> Int {
        var nums = nums

        var left = 0, right = nums.count - 1

        while true {
            let idx = partition(&nums, left, right)
            if idx + 1 == k {
                return nums[idx]
            } else if idx < k {
                left = idx + 1
            } else {
                right = idx - 1
            }
        }
    }
}

// decreasing nums
private func partition(_ nums: inout [Int], _ left: Int, _ right: Int) -> Int {
    let pivot = nums[right]
    var l = left

    for i in left ..< right {
        if nums[i] >= pivot {
            nums.swapAt(l, i)
            l += 1
        }
    }

    nums.swapAt(l, right)
    // print("nums:\(nums)\tpivot:\(pivot)\tl:\(l)")
    return l
}
